File: Ywy/Go/level1/NumberDescSort.go
package main

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
)

func main() {
	fmt.Println(solution(118372))
}

func solution(n int64) int64 {
	var str string = strconv.FormatInt(n, 10)
	arr := strings.Split(str, "")
	sort.Slice(arr, func(prev, next int) bool {
		return arr[prev] > arr[next]
	})

	str = ""

	for index := range arr {
		str += arr[index]
	}

	answer, _ := strconv.ParseInt(str, 0, 64)
	return answer
}

File: Ywy/JAVA/src/level1/NumberDescSort.java
package level1;

public class NumberDescSort {
	
	public long solution(final long n) {
		final char[] arr = String.valueOf(n).toCharArray();
		final int range = arr.length - 1;
		int index = 0;
		int jumpIndex = -1;
		char temp = 0;
		while (index < range) {
			temp = arr[index++];
			if (temp < arr[index]) {
				if (jumpIndex == -1) {
					jumpIndex = index;
				}
				arr[index - 1] = arr[index];
				arr[index] = temp;
				index -= index > 1 ? 2 : 0;
				continue;
			}
			
			if (jumpIndex > -1) {
				index = jumpIndex;
				jumpIndex = -1;
			}
		}
		return Long.parseLong(new String(arr));
	}
}
